lib/demo_web/controllers/user_auth.ex

```diff
@@ -0,0 +1,139 @@
+defmodule DemoWeb.UserAuth do
+  import Plug.Conn
+  import Phoenix.Controller
+
+  alias Demo.Accounts
+  alias DemoWeb.Router.Helpers, as: Routes
+
+  # Make the remember me cookie valid for 60 days.
+  # If you want bump or reduce this value, also change
+  # the token expiry itself in UserToken.
+  @max_age 60 * 60 * 24 * 60
+  @remember_me_cookie "user_remember_me"
+  @remember_me_options [sign: true, max_age: @max_age]
+
+  @doc """
+  Logs the user in.
+
+  It renews the session ID and clears the whole session
+  to avoid fixation attacks. See the renew_session
+  function to customize this behaviour.
+  """
+  def login_user(conn, user, params \\ %{}) do
```

Comment:
How about replacing the `params \\ %{}` argument with `remember_me \\ false` (+ relevant change in `UserSessionController`)? Params are passed only for sake of digging this boolean so that would make both `login_user/3` and `maybe_write_remember_me_cookie/3` easier to comprehend. Building on that, it would be easier to make a post-generator adjustment that'd make the remember me a default for cases when the field is not sent or when user logged in after registration - just change the default boolean here.

```suggestion
  def login_user(conn, user, remember_me \\ false) do
```

Reply:
I agree with that - also I wonder if it would be nice to have `login_user` and `logout_user` not actually include the `redirect` response, and only change sessions/cookies.

That way they can be used in different situations where you may not want to redirect as they are now (including in tests).

I guess that would mean either
  - the controllers that currently use `login_user` and `logout_user` manually redirect afterwards
  - or they use something like `login_user_and_redirect` and `logout_user_and_redirect` that does exactly what they're doing now, e.g.
```elixir
def logout_user_and_redirect(conn) do
  conn
  |> logout_user()
  |> redirect(to: "/")
end
```
(and similar for `login_user_and_redirect`)

Reply:
All of those improvements are interesting and I can see them being useful in some situations. However, I did my best to avoid "future guessing" this code and focus on the minimum that works for this PR. That's because for certain apps, what is here will definitely be enough.

The only cases where I did extract things out was when:

* To extract magic values, such as expiration intervals
* I had to do security remarks, such as the `renew_session` function in this module.